src/vms/pvm/etna-builder/spendHelper.ts

```diff
@@ -0,0 +1,333 @@
+import type { OutputOwners, TransferableOutput } from '../../../serializable';
+import { TransferableInput } from '../../../serializable';
+import type { Utxo } from '../../../serializable/avax/utxo';
+import { isTransferOut } from '../../../utils';
+import { bigIntMin } from '../../../utils/bigintMath';
+import { compareTransferableOutputs } from '../../../utils/sort';
+import type { Dimensions } from '../../common/fees/dimensions';
+import {
+  addDimensions,
+  createEmptyDimensions,
+  dimensionsToGas,
+} from '../../common/fees/dimensions';
+import { consolidateOutputs } from '../../utils/consolidateOutputs';
+import { getInputComplexity, getOutputComplexity } from '../txs/fee';
+
+export interface SpendHelperProps {
+  changeOutputs: readonly TransferableOutput[];
+  gasPrice: bigint;
+  initialComplexity: Dimensions;
+  inputs: readonly TransferableInput[];
+  shouldConsolidateOutputs: boolean;
+  stakeOutputs: readonly TransferableOutput[];
+  toBurn: Map<string, bigint>;
+  toStake: Map<string, bigint>;
+  weights: Dimensions;
+}
+
+/**
+ * The SpendHelper class assists in managing and processing the spending of assets,
+ * including handling complexities, gas prices, and various outputs and inputs.
+ *
+ * @class
+ */
+export class SpendHelper {
+  private readonly gasPrice: bigint;
+  private readonly initialComplexity: Dimensions;
+  private readonly shouldConsolidateOutputs: boolean;
+  private readonly toBurn: Map<string, bigint>;
+  private readonly toStake: Map<string, bigint>;
+  private readonly weights: Dimensions;
+
+  private changeOutputs: readonly TransferableOutput[];
+  private inputs: readonly TransferableInput[];
+  private inputComplexity: Dimensions = createEmptyDimensions();
+  private outputComplexity: Dimensions = createEmptyDimensions();
```

Comment:
Can we remove the `outputComplexity` slice of state? This is something that is very precarious to update because outputs are added/removed. There is no method to subtract complexity, so it's very easy to get this wrong. We can just recompute this complexity on demand. This will make the temporary output complexity thing much simpler... Similarly can we remove `inputComplexity`? Even though inputs are not removed, it's a derivable slice of state, we can just recompute it. In fact `getComplexity` does not even reference it

Reply:
Cleaned up. Artifacts from previous revisions.